src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/Common/DbConnectionOptions.cs

```diff
@@ -105,5 +107,83 @@ public bool ContainsKey(string keyword)
         {
             return _parsetable.ContainsKey(keyword);
         }
+
+        protected internal virtual string Expand()
+        {
+            return _usersConnectionString;
+        }
+
+        // SxS notes:
+        // * this method queries "DataDirectory" value from the current AppDomain.
+        //   This string is used for to replace "!DataDirectory!" values in the connection string, it is not considered as an "exposed resource".
+        // * This method uses GetFullPath to validate that root path is valid, the result is not exposed out.
+        internal static string ExpandDataDirectory(string keyword, string value)
+        {
+            string fullPath = null;
+            if ((null != value) && value.StartsWith(DataDirectory, StringComparison.OrdinalIgnoreCase))
+            {
+                // find the replacement path
+                object rootFolderObject = AppDomain.CurrentDomain.GetData("DataDirectory");
+                var rootFolderPath = (rootFolderObject as string);
+                if ((null != rootFolderObject) && (null == rootFolderPath))
+                {
+                    throw ADP.InvalidDataDirectory();
+                }
+                else if (string.IsNullOrEmpty(rootFolderPath))
+                {
+                    rootFolderPath = AppDomain.CurrentDomain.BaseDirectory ?? string.Empty;
+                }
+
+                // We don't know if rootFolderpath ends with '\', and we don't know if the given name starts with onw
```

Comment:
Lines 137-157 seem like a lot of effort for what can be consolidated down to a single Path.Join statement. (I see it's that way in the netfx code, but it seems like a simple improvement.)

Reply:
Actuslly, Path.Join does not do the job - I tried that! Fails on "C:\test" and "\attach.mdf" as as as I remember

Reply:
Fixed!